### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by DrunkWinter**

### CODE REVIEW

File: src/components/unlocker.js [hunk 1]
```diff
@@ -54,36 +55,31 @@ export function unlockPlayerResponse(playerResponse) {
...
-    function useProxy() {
+    if (playerResponse?.playabilityStatus?.status !== "OK")
         logger.info("Trying Unlock Method #2 (Account Proxy)");
-        return proxy.getPlayerFromAccountProxy(videoId, reason);
-    }
+        playerResponse = proxy.getPlayer(videoId, reason);
+
+    // Cache response for 10 seconds
+    // responseCache = { videoId, playerResponse };
+    // setTimeout(() => { responseCache = {} }, 10000);
```
Comment (DrunkWinter) [inline src/components/unlocker.js:73]: The cache has no obvious effect, why is this here in the first place? 
On top of that, the cache is useless after this function gets called multiple times, since we never clear the timeout.

File: src/main.js [hunk 2]
```diff
@@ -66,11 +66,11 @@ function onXhrOpenCalled(xhr, method, url) {
...
 
         // solve CORS errors by preventing YouTube from enabling the "withCredentials" option (not required for the proxy)
-        Object.defineProperty(xhr, "withCredentials", {
-            set: () => { },
-            get: () => false,
-        });
+        // Object.defineProperty(xhr, "withCredentials", {
+        //     set: () => { },
+        //     get: () => false,
+        // });
```
Comment (DrunkWinter) [inline src/main.js:72]: This doesn't seem like it has effect. Can you confirm?

File: src/components/unlocker.js (hunk 1 above)
Comment (zerodytrash) [reply src/components/unlocker.js:73]: there are often two unlocks. inital data and player request. to avoid two requests to youtube or my proxy, the cache was implemented.

File: src/main.js (hunk 2 above)
Comment (zerodytrash) [reply src/main.js:72]: this is only required if the account proxy was used. test it here: https://www.youtube.com/watch?v=gKVjWNv9V8Q
for me it is necessary.
image

Comment (DrunkWinter) [reply src/main.js:72]: The comment above it had me confused because it says "not required for proxy"

File: src/components/unlocker.js (hunk 1 above)
Comment (DrunkWinter) [reply src/components/unlocker.js:73]: Can you point me to where those unlocks happen for both? Because from what i can see is that it can be called only once for the video. `getUnlockedPlayerResponse` only gets called once by `unlockPlayerResponse` which can only get called in `main.js` also once.

File: src/main.js (hunk 2 above)
Comment (zerodytrash) [reply src/main.js:72]: yes, a bit complicated xd
This means that the withCredentials option is not needed for the proxy. Therefore setting the withCredentials option will be prevented by setting a descriptor that always returns false. This will also solve the CORS problems :)

Comment (DrunkWinter) [reply src/main.js]: Okay just remove the "not" from the comment which makes more sense ;)

File: src/components/unlocker.js (hunk 1 above)
Comment (zerodytrash) [reply src/components/unlocker.js:73]: Have tested it. Currently no second player request seems to be executed after the initial unlock. Only with the embedded player, there is another request after you click play.
https://kworb.net/youtube/video/Mi3GdCyzHCY.html

I think it would be good to keep the cache in case something unexpected happens (e.g. an endless loop during JSON interception).

Comment (DrunkWinter) [reply src/components/unlocker.js:73]: The embedded player does not work for me.

image

The cache does not work correctly at the moment and would not work at all for the endless loop case. So I have to fix it first.

Comment (DrunkWinter) [reply src/components/unlocker.js:73]: Can I just remove the setTimeout?

Comment (zerodytrash) [reply src/components/unlocker.js:73]: yes, you can replace it with something else which fulfills the same purpose.

> The embedded player does not work for me.

strange. have you tried with the current version?

Comment (DrunkWinter) [reply src/components/unlocker.js]: > yes, you can replace it with something else which fulfills the same purpose.

Removed the `SetTimeout` because it had no use. the cache already is meant to be reassigned if the video does not match cache id.
> > The embedded player does not work for me.
> 
> strange. have you tried with the current version?

Current version works fine now :)

**Comment by DrunkWinter** [comment]
This is ready.

**Comment by zerodytrash** [comment]
Nice strategy implementation! But there was a logic error with the `every` iteration. Only the first strategy was tried because `undefined` was returned.

**Comment by DrunkWinter** [comment]
> Nice strategy implementation! But there was a logic error with the `every` iteration. Only the first strategy was tried because `undefined` was returned.

Good catch!
